Send emails when reaching 90% and 100% of mobile user limit

minhaminha opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

A one-time email notice will now be sent out when the mobile worker count crosses the 90% and 100% threshold based on that domain's plan's mobile worker limit. 

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->

[Jira Ticket](https://dimagi-dev.atlassian.net/browse/SALES-26419)
[Specs](https://docs.google.com/document/d/1n3__riOHJTjYFxI6Vvq1JpBpJzXJv-ZH8jR7NCfovmM/edit) (also in ticket)

Emails will be sent regardless of if the mobile worker was added manually/individually or through the bulk upload action since they both use the new `check_and_send_limit_email` static function that checks whether the email should be sent at all and if so, which one, based on the given plan limit and user count from before and after the user creation action. `get_plan_and_user_count_by_domain` was separated out as its own function since I need to use it twice (before and after bulk upload action) to check whether the email should send or not. 

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->

Tested locally and on staging + QA tested. This doesn't change any data only gets it. 

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->

[QA Ticket](https://dimagi-dev.atlassian.net/browse/QA-4683)

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

minhaminha committed:
Send emails when user count reaches capacity or the 90% threshold

minhaminha committed:
Refactor into separate functions + send email when workers are uploaded via bulk upload

minhaminha committed:
Fix email text to match specs

minhaminha committed:
Modify logic for getting the right rate

mjriley commented on corehq/apps/users/views/mobile/users.py:
```diff
@@ -756,11 +763,61 @@ def create_mobile_worker(self, in_data):
+            if rate.feature.feature_type == 'User':
+                user_rate = rate
+                break
+        return {
+            'plan_limit': user_rate.monthly_limit,
+            'user_count': FeatureUsageCalculator(user_rate, domain).get_usage()
+        }
```
I notice that all usages of this essentially boil down to:
```
data = get_plan_and_user_count_by_domain(...)`
do_something(data['plan_limit'], data['user_count'])
```

I appreciate the readability of an object with specific keys, but I feel like that readability is mostly offset by kind of pushing the user into using a generic `data` variable. Javascript has a clean solution to this, as you can just write:
`const {plan_limit, user_count} = get_plan_and_user_count_by_domain(...)`. There is some discussion of the ways we could get Javascript-like functionality [here](https://stackoverflow.com/questions/54785148/destructuring-dicts-and-objects-in-python), but I wonder if it's just more Pythonic to make the function itself slightly less readable by returning a tuple instead of a dict, i.e.:
` return (user_rate.monthly_limit, FeatureUsageCalculator(user_rate, domain).get_usage()`
so that it enables just doing:
` plan_limit, user_count = get_plan_and_user_count_by_domain(...)`

mjriley commented on corehq/apps/users/views/mobile/users.py:
```diff
@@ -756,11 +763,61 @@ def create_mobile_worker(self, in_data):
+        else:
+            return None
+
+        users = WebUser.by_domain(domain)
+        recipients = []
+        for user in users:
+            if user.role_label(domain) == 'Billing Admin' or user.role_label(domain) == 'Admin':
```
This logic seems separate to what this method is doing, and it's likely re-usable. I'd imagine other places in our code may want to get a list of billing administrators in the future, and in that case, we wouldn't want to duplicate the logic that both billing admin and admins are considered billing administrators.  _corehq.apps.users.models.role_utils_ also exposes a `BILLING_ADMIN` constant that we probably want to use in place of the string here

minhaminha replied:
Made this a classmethod of WebUser since there's already a `get_admins_by_domain` type of function

mjriley commented on corehq/apps/users/views/mobile/users.py:
```diff
@@ -756,11 +763,61 @@ def create_mobile_worker(self, in_data):
+        else:
+            return None
+
+        users = WebUser.by_domain(domain)
+        recipients = []
+        for user in users:
+            if user.role_label(domain) == 'Billing Admin' or user.role_label(domain) == 'Admin':
```
`role_label()` appears to make database calls. Consider storing it in a variable to avoid making multiple calls

mjriley commented on corehq/apps/users/views/mobile/users.py:
```diff
@@ -756,11 +763,61 @@ def create_mobile_worker(self, in_data):
+            subject = _("User count has reached the Plan limit for {}").format(domain)
+        else:
+            subject = _("User count has reached 90% of the Plan limit for ({})").format(domain)
+        send_HTML_email(
```
I see two calls for `check_and_send_limit_email` in tasks modules, and one from `create_mobile_worker`. It's probably fine to call this synchronous version of email for the tasks, but are we sure it's safe to do so for how `create_mobile_worker` is used? Should we consider `send_html_email_async` instead?

minhaminha replied:
It's probably fine to keep it the way it was but `send_html_email_async` retries sending it multiple times which seems important for an email that's supposed to send exactly twice.

mjriley commented on corehq/apps/users/templates/users/email/user_limit_notice.html:
```diff
@@ -0,0 +1,15 @@
+    Please note that you have reached the user limit included in your plan.
+    Your user limit represents all mobile users that are not currently deactivated
+    (see <a href="{{ url }}">here</a> for more). If you exceed the number of mobile users
+    included in your plan, you will be charged as per the terms of plan.
```
> per the terms of plan.

the plan?

mjriley commented on corehq/apps/users/templates/users/email/user_limit_notice.html:
```diff
@@ -0,0 +1,15 @@
+{% if at_capacity %}
```
Would it be helpful to include the actual numbers vs capacity here? Specifically for the 90% message, I feel it would be useful to know that they could create another 5 or 500 users, for example

RynhardPietersen replied:
I think you can take out the entire if statement here by using something along the lines of
```
"You have reached {created_users} out of {maximum_users} allowed in your plan. Your user limit represents all mobile users that are not currently deactivated
    (see <a href="{{ url }}">here</a> for more). If you exceed the number of mobile users
    included in your plan, you will be charged as per the terms of plan."
```

minhaminha replied:
The text is lifted directly from the specs and I'm not sure if I'm allowed to mess with it without approval.

RynhardPietersen replied:
Yeah that makes sense. Would it make sense to speak to the creator of the spec about at least including actual numbers here instead?

mjriley commented on corehq/apps/users/views/mobile/users.py:
```diff
@@ -756,11 +763,61 @@ def create_mobile_worker(self, in_data):
             'user_id': couch_user.userID,
         }
 
+    @staticmethod
+    def get_plan_and_user_count_by_domain(domain):
```
I don't think this belongs in a view module. Re-usable computation seems more like a model-level concern, and I think having our modules import other view code tends to lead to circular dependency issues.

minhaminha replied:
Moved to be a classmethod of the Subscription class

mjriley reviewed (changes requested):
I don't have access to the spec, so apologies if some of these issues have already been settled there. That said, where are the tests? Presumably it's important that these emails are sent out when the thresholds are met -- there should be tests proving that we send out those emails, that the emails are the correct ones, and that we don't send out repeat emails, if that is the intended behavior.

RynhardPietersen commented on corehq/apps/users/views/mobile/users.py:
```diff
@@ -756,11 +763,61 @@ def create_mobile_worker(self, in_data):
+                'url': ADDITIONAL_USERS_PRICING
+            }),
+        )
+        return None
```
Is this return doing anything?

minhaminha replied:
Nope but I think it's good practice to make clear a function won't return anything?

mjriley replied:
I don't think this aligns with our other code. I would disagree that this is good practice, as that would mean putting essentially a no-op return statement at the bottom of every function. Seeing `return None` rather than `return` (which I believe is the intent) confused me a bit, as it implies that this function _might_ return something, and that its callers _might_ care about its return value. I'd advocate to modify the earlier `return None`s to simple `return` statements, and leave off the `return None` at the end

minhaminha committed:
PR feedback refactoring

minhaminha commented on corehq/apps/users/views/mobile/users.py:
```diff
@@ -756,11 +763,61 @@ def create_mobile_worker(self, in_data):
+        for rate in plan_version.feature_rates.all():
+            if rate.feature.feature_type == 'User':
+                user_rate = rate
+                break
```
Adopted the suggestion! and I'm pretty certain all plans do have the User feature type (and so a plan limit) since you can create mobile workers with any plan

minhaminha commented on corehq/apps/users/views/mobile/users.py:
```diff
@@ -756,11 +763,61 @@ def create_mobile_worker(self, in_data):
+
+    @staticmethod
+    def check_and_send_limit_email(domain, plan_limit, user_count, prev_count=None):
+        if plan_limit == -1:
```
saved as a variable to avoid confusion but Enterprise plans allow for unlimited mobile workers and so it sets the plan limit at -1

minhaminha commented on corehq/apps/users/views/mobile/users.py:
```diff
@@ -756,11 +763,61 @@ def create_mobile_worker(self, in_data):
+            return None
+
+        if prev_count is None:
+            prev_count = user_count - 1
```
Sorry for the confusion! There's only two ways to add mobile workers: individually using the Create Mobile Workers modal or through bulk uploads. If there wasn't a `prev_count` it meant that it was created through the latter means and so it'd always be 1 less than what it currently is (since that info is retrieved after the user is fully created). I made this more clear by making `prev_count` non optional and explicitly passing `user_count-1` as the input.

minhaminha commented on corehq/apps/users/views/mobile/users.py:
```diff
@@ -756,11 +763,61 @@ def create_mobile_worker(self, in_data):
+        elif plan_limit > user_count >= (0.9 * plan_limit) > prev_count:
+            at_capacity = False
+        else:
+            return None
```
From the looks of the specs that is the intended behavior but I can ask Dev to confirm.

minhaminha commented:
Sorry for the delay @mjriley but I think I've addressed all your concerns (thanks by the way for the thorough review!!) except the tests which I'm still working on.

RynhardPietersen commented:
I like the changes! I'm not sure about the name of the commit though. `Pr feedback refactoring` makes sense in the context of this pr but if and when this pr is eventually merged it would look pretty vague if someone is trying to revert to a previous commit.

minhaminha committed:
Add test to ensure emails are being sent

minhaminha committed:
Add user count to email body

mjriley commented on corehq/apps/users/views/mobile/users.py:
```diff
@@ -756,11 +761,47 @@ def create_mobile_worker(self, in_data):
+        billing_admins = [admin.username for admin in WebUser.get_billing_admins_by_domain(domain)]
+        admins = [admin.username for admin in WebUser.get_admins_by_domain(domain)]
+
+        if at_capacity:
+            subject = _("User count has reached the Plan limit for {}").format(domain)
+        else:
+            subject = _("User count has reached 90% of the Plan limit for {}").format(domain)
+        send_html_email_async(
+            subject,
+            admins + billing_admins,
```
I don't know how well this will handle translations if the admins don't share the same language as whoever called the `create_mobile_worker` function. `subject` will be translated to whatever language the caller has set. If this is a concern, you could look at something like [the report notification code](https://github.com/dimagi/commcare-hq/blob/7ac73b311c325fa3824dc8b2d4c32e3e7b784526/corehq/apps/saved_reports/models.py#L739-L741), which is one example I'm aware of where we localize the emails to the recipient.

minhaminha replied:
The function you linked (or more specifically `recipients_by_language`) sounds like something that could be reworked to be more flexibly applied to different areas so I'll make a note to come back to this as it's own thing. But I don't really think different translations will be a problem here since the emails in this scenario are limited to a singular project's admins (who I presume work together).

mjriley replied:
I'm not sure of the scope of the problem. I overlooked that both the `create_mobile_worker` function and `import_users` will hit these translations. While it's likely that the language for the user who triggers `create_mobile_worker` is shared by the admins, I'm less sure that tasks inherit the language settings of their caller. If not, then you could have a project where users and admins use something like Swahili receiving an email in English. [Possibly relevant celery link?](https://stackoverflow.com/questions/7199256/translation-in-celery-tasks#comment58178158_7208163). I don't think this should block this work, but its something to keep in mind.

minhaminha replied:
Yea I think you’re right needing to explicitly reference the caller/user’s language in order for the translations to take effect but I’m having a hard time finding other examples in the code, since I assume there’s at least one or two other instances where tasks trigger emails with translated strings. But in the interest of time I’m gonna lump this together with the previous note I left above and come back to both when I have more time!

mjriley commented on corehq/apps/users/tests/test_email_events.py:
```diff
@@ -36,3 +42,51 @@ def test_bounced_event(self):
+        plan_limit = 100
+
+        # email should not send if user count is still below the 90% threshold
+        prev_user_count = 75
+        post_user_count = 85
```
Because off-by-one errors are so common, I generally like to test right on the boundaries. Since its greater-than-or-equal comparisons, that would mean no emails up to 89 users, and ensuring an email is sent on exactly 90 users.

mjriley commented on corehq/apps/users/models.py:
```diff
@@ -2402,6 +2402,14 @@ def get_admins_by_domain(cls, domain):
 
+    @classmethod
+    def get_billing_admins_by_domain(cls, domain):
+        from corehq.apps.users.role_utils import UserRolePresets
```
If this import doesn't cause a circular import, I'd move it outside of this function to the top of the module. I don't know that we have standards around this, but my impression was that we're only using local imports to avoid circular imports because they're a code smell.

minhaminha replied:
yep this was because of circular import issue

mjriley commented on corehq/apps/user_importer/tasks.py:
```diff
@@ -102,6 +103,9 @@ def import_users(domain, user_specs, group_specs, upload_user_id, upload_record_
     def _update_progress(value, start=0):
         DownloadBase.set_progress(task, start + value, total)
 
+    from corehq.apps.users.views.mobile.users import MobileWorkerListView
```
Up to you, but I'd consider whether `check_and_send_limit_email` belongs inside a view module. I don't think it's functionality inherently fits with a view. It feels a bit odd to see this importer task forced to import from a view module to get this functionality

minhaminha replied:
Do you mind explaining why this is odd? I'm a little confused by what you mean by its "inherent functionality", though I do see that it would also fit neatly as a separate task in tasks.py. But then the view function would have to import it as well, which I guess could avoid the circular import issue it's having now.

mjriley replied:
I think it was explained to me that in architectures like Model-View-Controller (MVC) or N-Tier, each tier only knows about the tiers beneath it. So a View can know about and use Model code, but not the other way around (you can see how this avoids circular dependencies). I've traditionally viewed View code as the end tier, where no code knows about it except other View code. So that's why its weird to see behavior that lower levels might use, put into the highest level.

minhaminha replied:
that's a helpful way of looking at it, thanks! Moved to models.py

mjriley reviewed (changes requested):
I'd like to see the tests split apart, but other than that, the code looks good to go. Most of my comments are minor or suggestions.

minhaminha commented on corehq/apps/users/tests/test_email_events.py:
```diff
@@ -36,3 +42,51 @@ def test_bounced_event(self):
+        self.billing_admin_user.delete(self.domain.name, None)
+        self.domain.delete()
+
+    def test_limit_emails(self):
```
I'm tempted to roll the last two into the respective cases that also test whether they send at the right triggers (at/past threshold and at/past the maximum) because if they were their own tests, I'd have to do that again + make sure that it doesn't send while remaining above the 90% or 100% threshold.  Would that be an issue of cramming too many things into a single test again?

minhaminha replied:
Also, in general, how granular should these tests be? I haven't been great about adding tests with new features so I'm trying to get a feel for how much is enough, going forward.

mjriley replied:
Ideally, every condition is its own test case. Even if things are related, it doesn't mean they aren't implemented in a way that an error in one part of the condition implies success or failure in the other part of the condition. Putting them in the same test case falls back into the same problems with stopping at the first failure and inheriting state. If the code strongly makes you want to avoid re-creating data for separate test cases, that can be a good indication that refactoring is necessary to make the code easier to test.

As a shortcut, you might hear someone say that a test should only have one assertion. That is incorrect -- a test should only be testing one thing, and sometimes that one thing requires multiple assertions (i.e that all items in an array match).  A useful shortcut, however, is seeing multiple assertions _with test setup between them_. That is two tests masquerading as one. 

If you're not familiar with it, check out the [AAA (Arrange, Act, Assert) pattern](https://robertmarshall.dev/blog/arrange-act-and-assert-pattern-the-three-as-of-unit-testing/).

I might be missing something, but in this case, I actually think the code seems easy enough to test? Because your function takes in all the relevant variables (`plan_limit`, `user_count`, and `prev_count`), you don't need to trigger a previous email before you can verify that a second call will skip the email -- you just have to make sure `prev_count` is above the threshold you're testing.

minhaminha committed:
Break up test, add more tests, move main func to models.py

mjriley commented on corehq/apps/users/models.py:
```diff
@@ -3054,3 +3062,38 @@ def get_deactivate_after_date(cls, domain, user_id):
+        subject = _("User count has reached 90% of the Plan limit for {}").format(domain)
+    send_html_email_async(
+        subject,
+        admins + billing_admins,
```
There's an edge case here, where admins and billing_admins are not necessarily distinct. For example, a 'global admin' (superuser) can be invited to a domain as a billing admin, in which case he'd appear in both lists. If that were true, he'd likely receive two emails. Would probably be safer to union the two as a set

mjriley commented on corehq/apps/users/tests/test_email_events.py:
```diff
@@ -36,3 +40,72 @@ def test_bounced_event(self):
+        self.mock_get_billing_admins.return_value = [WebUser(username=self.recipients[1])]
+        self.addCleanup(billing_admin_patcher.stop)
+
+    def tearDown(self):
+        super().tearDown()
```
If this method isn't doing anything but calling the parent, it should be removed. The testing framework will handle calling the `tearDown` method on its own.

minhaminha replied:
ah, I didn't know that tearDown is called automatically. Noted!

mjriley reviewed (approved):
Thanks for the changes! Looks great!

minhaminha committed:
Remove tearDown, convert recipient list to set

minhaminha commented:
> Thanks for the changes! Looks great!

Thank YOU for your detailed responses!! At this point you should be a co-author for this PR :)

minhaminha committed:
Stickler fix (removing unnecessary import)
